src/platform_impl/linux/wayland/seat/pointer/mod.rs

```diff
@@ -467,9 +469,107 @@ impl Dispatch<WpCursorShapeManagerV1, GlobalData, WinitState> for SeatState {
+    fn event(
+        _state: &mut WinitState,
+        _proxy: &ZwpPointerGesturesV1,
+        _event: <ZwpPointerGesturesV1 as wayland_client::Proxy>::Event,
+        _data: &GlobalData,
+        _conn: &Connection,
+        _qhandle: &QueueHandle<WinitState>,
+    ) {
+        unreachable!("zwp_pointer_gestures_v1 has no events")
+    }
+}
+
+impl Dispatch<ZwpPointerGesturePinchV1, GlobalData, WinitState> for PointerGesturesState {
+    fn event(
+        state: &mut WinitState,
+        _proxy: &ZwpPointerGesturePinchV1,
+        event: <ZwpPointerGesturePinchV1 as wayland_client::Proxy>::Event,
+        _data: &GlobalData,
+        _conn: &Connection,
+        _qhandle: &QueueHandle<WinitState>,
+    ) {
+        let device_id = crate::event::DeviceId(crate::platform_impl::DeviceId::Wayland(DeviceId));
+        let (phase, pan_delta, scale_delta, rotation_delta) = match event {
+            Event::Begin { time: _, serial: _, surface, fingers } => {
+                if fingers != 2 {
+                    // We only support two fingers for now.
+                    return;
+                }
+                // The parent surface.
+                let parent_surface = match surface.data::<SurfaceData>() {
+                    Some(data) => data.parent_surface().unwrap_or(&surface),
+                    None => return,
+                };
+
+                let pan_delta = PhysicalPosition::new(0., 0.);
+                state.window_id = wayland::make_wid(parent_surface);
+                state.previous_scale = 1.;
+
+                (TouchPhase::Started, pan_delta, 0., 0.)
+            },
+            Event::Update { time: _, dx, dy, scale, rotation } => {
+                let pan_delta = PhysicalPosition::new(dx as f32, dy as f32);
+                let scale_delta = scale - state.previous_scale;
+                state.previous_scale = scale;
+                (TouchPhase::Moved, pan_delta, scale_delta, -rotation as f32)
+            },
+            Event::End { time: _, serial: _, cancelled } => {
+                let pan_delta = PhysicalPosition::new(0., 0.);
+                state.previous_scale = 1.;
+                (
+                    if cancelled == 0 { TouchPhase::Ended } else { TouchPhase::Cancelled },
+                    pan_delta,
+                    0.,
+                    0.,
+                )
+            },
+            _ => unreachable!("Unknown event {event:?}"),
+        };
+        state.events_sink.push_window_event(
+            WindowEvent::PanGesture { device_id, delta: pan_delta, phase },
+            state.window_id,
+        );
+        state.events_sink.push_window_event(
+            WindowEvent::PinchGesture { device_id, delta: scale_delta, phase },
+            state.window_id,
+        );
+        state.events_sink.push_window_event(
+            WindowEvent::RotationGesture { device_id, delta: rotation_delta, phase },
+            state.window_id,
+        );
```

Comment:
You should only send updates if there was a change compared to previous state, so if there was not pinch, but just rotation, only rotation should be send.